peps/pep-0782.rst

```diff
@@ -0,0 +1,371 @@
+
+.. c:function:: void* PyBytesWriter_GrowAndUpdatePointer(PyBytesWriter *writer, Py_ssize_t size, void *buf)
+
+   Similar to :c:func:`PyBytesWriter_Grow`, but update also the *buf*
+   pointer.
+
+
+Overallocation
+--------------
+
+:c:func:`PyBytesWriter_Resize` overallocates the internal buffer to
+reduce the number of ``realloc()`` calls and so reduce memory copies.
+
+
+Thread safety
+-------------
+
+The API is not thread safe: a writer should only be used by a single
+thread at the same time.
+
+
+Examples
+--------
+
+High-level API
+^^^^^^^^^^^^^^
+
+Create the string ``"Hello World!"``::
+
+    PyObject* hello_world(void)
+    {
+        PyBytesWriter *writer = PyBytesWriter_Create(0);
+        if (writer == NULL) {
+            goto error;
+        }
+        if (PyBytesWriter_WriteBytes(writer, "Hello", -1) < 0) {
+            goto error;
+        }
+        if (PyBytesWriter_Format(writer, " %s!", "World") < 0) {
+            goto error;
+        }
+        return PyBytesWriter_Finish(writer);
+
+    error:
+        PyBytesWriter_Discard(writer);
+        return NULL;
+    }
+
+
+Create the bytes string "abc"
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Example creating the string ``"abc"``, with a fixed size of 3 bytes::
+
+    PyObject* create_abc(void)
+    {
+        PyBytesWriter *writer = PyBytesWriter_Create(3);
+        if (writer == NULL) {
+            return NULL;
+        }
+
+        char *str = PyBytesWriter_GetData(writer);
+        memcpy(str, "abc", 3);
+        return PyBytesWriter_Finish(writer);
+    }
+
+GrowAndUpdatePointer() example
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Example using a pointer to write bytes and to track the written size.
+
+Create the string ``"Hello World"``::
+
+    PyObject* grow_example(void)
+    {
+        // Allocate 10 bytes
+        PyBytesWriter *writer = PyBytesWriter_Create(10);
+        if (writer == NULL) {
+            return NULL;
+        }
+
+        // Write some bytes
+        char *buf = PyBytesWriter_GetData(writer);
+        memcpy(buf, "Hello ", strlen("Hello "));
+        buf += strlen("Hello ");
+
+        // Allocate 10 more bytes
+        buf = PyBytesWriter_GrowAndUpdatePointer(writer, 10, buf);
+        if (buf == NULL) {
+            PyBytesWriter_Discard(writer);
+            return NULL;
+        }
+
+        // Write more bytes
+        memcpy(buf, "World", strlen("World"));
+        buf += strlen("World");
+
+        // Truncate the string at 'buf' position
+        // and create a bytes object
+        return PyBytesWriter_FinishWithPointer(writer, buf);
+    }
+
+
+Implementation
+==============
+
+* `Pull request gh-131681 <https://github.com/python/cpython/pull/131681>`__.
+
+
+Backwards Compatibility
+=======================
+
+There is no impact on the backward compatibility, only new APIs are
+added.
+
+
+Projects using _PyBytes_Resize()
+================================
+
+A code search on PyPI top 8,000 projects finds 41 projects using
+``_PyBytes_Resize``:
```

Comment:
What is this section for? Perhaps it should go into Motivation or Rationale if you want to demonstrate a demand for the new API.

Reply:
I just removed the section.